File: Python/DMOJ/interpretive_art.py
from copy import deepcopy
#*If hard to debug in C++, will do it in Python. 

def custom_sort(arr, l, h):
    #*l and h include the index. 
    temp = []
    for i in range(l, h+1):
        temp.append(arr[i])
    temp.sort()
    curr = 0
    for i in range(l,h+1):
        arr[i] = temp[curr]
        curr+=1
    return arr

n = input()
n = int(n)

a_num = 0
b_num = 0

a = input().strip().split(" ")
b = input().strip().split(" ")
for i in range(n):
    if a[i] == '0':
        a_num+=1

for i in range(n):
    if b[i] == '0':
        b_num+=1

if a_num != b_num:
    print(-1)
else:
    queue = [[a, []]]
    while len(queue) > 0:
        curr = queue.pop()
        if (curr[0]==b):
            print(len(curr[1]))
            for i in range(len(curr[1])):
                print(f"{curr[1][i][0]} {curr[1][i][1]}")
            break
        else:
            for i in range(len(curr[0])):
                for j in range(i, len(curr[0]), 1):
                    temp = deepcopy(curr)
                    custom_sort(temp[0], i, j)
                    temp[1].append([i, j])
                    #*Why do I just don't seem to reach the correct answer?
                    queue.append(temp)
                

File: C++/DMOJ cpp/interpretive_art.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
using namespace std;

vector <char> a;
vector <char> b;

void showVector(vector <char> arr){
    for (long long i=0;i<arr.size();i++){
        cout << arr[i] <<  " ";
    }
    cout << "\n";
}
int main(){
    //*Kind of similar to codejam pancake question. 
    //*I think the only case there is no solution is when there are different number of 0s and 1s. 

    long long n;
    long long a_num0 = 0;
    long long b_num0 = 0;
    cin >> n;
    char temp;
    for (long long i=0;i<n;i++){
        cin >> temp;
        a.push_back(temp);
        if (temp=='0'){
            a_num0++;
        }
    }

    for (long long i=0;i<n;i++){
        cin >> temp;
        b.push_back(temp);
        if (temp=='0'){
            b_num0++;
        }
    }

    if (a_num0 != b_num0){
        cout << -1;
    }
    else{
        //*We could just try naive BFS and hope it passes first batch. 
        vector <pair <vector <char>, vector<vector<long long>>>> queue = {{a, {}}};
        pair <vector <char>, vector<vector<long long>>> curr;
        pair <vector <char>, vector<vector<long long>>> temp;
        //map <vector <char>, vector <char>> came_from;
        //*We need to store where it came from so something like A*. 
        //!Not came from but rather the order of strikes. 

        //*I can store in pair. 
        while (queue.size() > 0){
            curr = queue[0];
            //!cout << curr.first[1]; This indexing works. 
            queue.erase(queue.begin());
            if (curr.first==b){
                //*Just print the remaining. 
                cout << curr.second.size() << "\n";
                for (long long i=0;i<curr.second.size();i++){
                    cout << curr.second[i][0] << " " << curr.second[i][i] << "\n";
                }
                
                break;
                
            }
            else{
                //*Have to copy for every movement. 
                //!Kind of confusing but I think this problem uses 1 indexing. 
                //!The program just somehow stops after the first 5 moves. 
                for (long long i=0;i<curr.first.size();i++){
                    //*No point sorting the same place twice. 
                    //!The sorting index is not inclusive. I will add 1 at the end to cope for that. 
                    for (long long j=i;j<curr.first.size();j++){
                        //*Copy the pair then sort. 
                        temp = curr;
                        sort(temp.first.begin()+i, temp.first.begin()+j+1);
                        //*I should show the vector since cannot access from debug. 
                        showVector(temp.first);
                        //!2 is correct but how can you just swap 00, 56 and reach the answer???
                        temp.second.push_back({i, j});
                        queue.push_back(temp);
                    }
                }
            }
            cout << "running";
        }
        cout << "done";
    }
}